### PR TITLE
Pre-merge tests

### PR DESCRIPTION
**Pull request by TrevorBenson**
Tests to validate changes to scripts on multiple OS's.

### CODE REVIEW

**Comment by TrevorBenson** [comment]
@rdlrt @Scitz0 In case you were interested in some additional pre-merge tests this should be useful. For instance when a "helpful soul" submits a PR which has a typo in an RPM package name (:point_right: :man_facepalming: :point_left: ). 

Currently only includes testing prereqs.sh and cabal-build-all.sh for stock and w/ system libsodium. I only ran the w/ system libsodium myself, but added the w/o so it gets tested as well. Obviously rust, hw wallet or any combination thereof could be setup as well.

File: .github/workflows/premerge_amazonlinux2_cabal-build-all.sh_-l.yml
Comment (rdlrt) [inline .github/workflows/premerge_amazonlinux2_cabal-build-all.sh_-l.yml:19]: Can we make this a second job within same action file (one with system libsodium based cabal-build-all.sh and second job with IO fork)? Helps reduce workflow quota from github, tho we prolly wouldnt have frequent changes to these files, but atleast combining tests for libsodium directly divides the workflows by 2 (other alternate would be using all OS against a specific test within same action file), either of those should be alright

File: .github/workflows/premerge_amazonlinux2_prereq.sh_-l.yml
Comment (rdlrt) [inline .github/workflows/premerge_amazonlinux2_prereq.sh_-l.yml:19]: Same as previous comment, but for prereqs.sh

File: files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile
Comment (rdlrt) [inline files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile:14]: I am not sure if I fully understand the use of master branch (I have not tested it on my end yet - just reading, so excuse me if I misunderstood) - we'd want to use the branch from the commit before the merge right (also, the merge would likely be targetting alpha branch - as no PR done directly against master will be sustained, they get overwritten by alpha)?

If above query is valid, the same would prolly apply to every file in this folder

**Review by rdlrt — COMMENTED** [review]
Thanks, that's a pretty good start indeed - and would be nice addition.
But just a few bits that we might want to check (added as comments).

**Comment by TrevorBenson** [comment]
I forgot this PR still needs to target using the prereqs.sh from the PR branch as well as call the PR branch for the correct cabal-build-all.sh.

File: files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile
Comment (TrevorBenson) [reply files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile:14]: Commented just a bit ago I forgot to implement using the PR branch, vs. master. Will fix this up

File: .github/workflows/premerge_amazonlinux2_cabal-build-all.sh_-l.yml
Comment (TrevorBenson) [reply .github/workflows/premerge_amazonlinux2_cabal-build-all.sh_-l.yml:19]: Sure. I had them split to increase observability by running everything in parallel, but I could easily see an active repository exceeding the pool of free Action minutes per month. I'l streamline it per OS. Also in another week CentOS 8 could be dropped due to EOL. The RockyLinux8 tests have produced identical results to all CentOS8 build testing.

The AmazonLinux2 image appears to be based on an fc35 kernel but without dnf. The `--allowerasing` "yum" subcommand is not really part of yum, but the `alias yum='dnf'` from el8 distro's hides that. This blocks the package installation on AL2. I presume the same goes for any el7 based distro currently. I don't use el7 anymore, or AL2 for anything. If the quantity of users does not warrant supporting it with prereqs.sh, that test could also be dropped.

Comment (TrevorBenson) [reply .github/workflows/premerge_amazonlinux2_cabal-build-all.sh_-l.yml:19]: Actually I guess its just 20 concurrent jobs in the queue since its a public repository, and the 2000 minutes per month limit would be only if the repo was private?

In any case I will work to reduce the concurrency and the OS's that you are not concerned with testing.

Comment (rdlrt) [reply .github/workflows/premerge_amazonlinux2_cabal-build-all.sh_-l.yml:19]: I agree on CentOS not being essential (especially since you cover Rocky), about AL2 - yes, there are a few consumers as per issues seen previously, so would be good to retain.
Thanks

Comment (TrevorBenson) [reply .github/workflows/premerge_amazonlinux2_cabal-build-all.sh_-l.yml:19]: OK. I'll submit a separate PR to add support for AL2 in prereqs.sh.

**Comment by TrevorBenson** [comment]
@rdlrt I went ahead and removed the CentOS tests and workflows. GITHUB_REF_NAME is passed in as build argument BRANCH. Any downloads of `prereqs.sh` come from `${BRANCH}` and it is always called with `-b ${BRANCH}` regardless of which flags we are testing it with (-l, etc.)

The original layout took just over 1 hour for pre-merge testing in parallel. Since we are chaining workflows, and there is no storage cap for the public free accounts, I added ghcr.io so we store the image from the initial prerequisites tests to be pulled by the cabal tests. Chained, even with ghcr.io my estimate is premerge testing will take around 4 hours +/- 30 minutes.

For the `cardano-node-latest.txt`, `cabal.project.local` and when kicking off the next workflow for pre-merge testing, I was calling all from `master` branch. 

- The workflow I think makes sense to keep pulling from master, or possibly alpha. But not `${{GITHUB_REF_NAME}}` in case an edit to the workflow actually breaks the per-merge tests.
- The `cabal.project.local` I could see pulling from the `${{GITHUB_REF_NAME}}` in case testing changes in the actual file. Otherwise either `master` or `alpha` seem acceptable.
- The `cardano-node-latest.txt` I think would probably be best to keep pulling from master as thats where I believe the action is storing it?

I'm in no rush to get it merged, so feel free to think through and make more suggestions until it's as you'd like.

File: .github/workflows/premerge_amazonlinux2_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_amazonlinux2_prereqs.sh.yml:9]:
```suggestion
      - scripts/cnode-helper-scripts/prereqs.sh
      - scripts/cnode-helper-scripts/cabal-build-all.sh
      - files/cabal.project.local

```

File: .github/workflows/premerge_debian_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_debian_prereqs.sh.yml:9]:
```suggestion
      - scripts/cnode-helper-scripts/prereqs.sh
      - scripts/cnode-helper-scripts/cabal-build-all.sh
      - files/cabal.project.local

```

File: .github/workflows/premerge_rockylinux8_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_rockylinux8_prereqs.sh.yml:9]:
```suggestion
      - scripts/cnode-helper-scripts/prereqs.sh
      - scripts/cnode-helper-scripts/cabal-build-all.sh
      - files/cabal.project.local

```

File: .github/workflows/premerge_ubuntu20_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_ubuntu20_prereqs.sh.yml:9]:
```suggestion
      - scripts/cnode-helper-scripts/prereqs.sh
      - scripts/cnode-helper-scripts/cabal-build-all.sh
      - files/cabal.project.local

```

File: files/tests/pre-merge/amazonlinux2-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/amazonlinux2-cabal.containerfile:21]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

File: files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile:21]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

File: files/tests/pre-merge/debian-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/debian-cabal.containerfile:22]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

File: files/tests/pre-merge/ubuntu20-cabal_-l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/ubuntu20-cabal_-l.containerfile:22]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

File: files/tests/pre-merge/rockylinux8-cabal_-l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/rockylinux8-cabal_-l.containerfile:21]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

File: files/tests/pre-merge/debian-cabal_-l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/debian-cabal_-l.containerfile:22]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

File: files/tests/pre-merge/ubuntu20-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/ubuntu20-cabal.containerfile:22]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

File: files/tests/pre-merge/rockylinux8-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/rockylinux8-cabal.containerfile:21]:
```suggestion
    curl -o cabal.project.local https://raw.githubusercontent.com/cardano-community/guild-operators/"${BRANCH}"/files/cabal.project.local &&\
```

**Comment by TrevorBenson** [comment]
> EDIT: Thinking this through and seeing default on alpha, and that PR's branch from alpha, I'm thinking the three bullets above should point to alpha instead of master. Except maybe `cabal.project.local` which could point to GITHUB_REF_NAME for the PR branch, in case testing new cabal param values within the submitted PR is the intended goal.

I'll bundle the suggested changes I added for reference and apply them if this is what we want, otherwise I'll clean them up and make it match. Sorry for all the extra noise in the PR comments, just wanted to keep track of everything before I went to bed for the night.

File: .github/workflows/premerge_amazonlinux2_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_amazonlinux2_prereqs.sh.yml:46]:
```suggestion
        curl -XPOST -u "${{ secrets.REPO_SCOPED_USER}}:${{secrets.REPO_SCOPED_TOKEN }}" -H "Accept: application/vnd.github.everest-preview+json" -H "Content-Type: application/json" https://api.github.com/repos/cardano-community/guild-operators/actions/workflows/premerge_amazonlinux2_cabal-build-all.sh.yml/dispatches --data '{"ref": "alpha"}'
```

File: .github/workflows/premerge_debian_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_debian_prereqs.sh.yml:46]:
```suggestion
        curl -XPOST -u "${{ secrets.REPO_SCOPED_USER}}:${{secrets.REPO_SCOPED_TOKEN }}" -H "Accept: application/vnd.github.everest-preview+json" -H "Content-Type: application/json" https://api.github.com/repos/cardano-community/guild-operators/actions/workflows/premerge_debian_cabal-build-all.sh.yml/dispatches --data '{"ref": "alpha"}'
```

File: .github/workflows/premerge_rockylinux8_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_rockylinux8_prereqs.sh.yml:46]:
```suggestion
        curl -XPOST -u "${{ secrets.REPO_SCOPED_USER}}:${{secrets.REPO_SCOPED_TOKEN }}" -H "Accept: application/vnd.github.everest-preview+json" -H "Content-Type: application/json" https://api.github.com/repos/cardano-community/guild-operators/actions/workflows/premerge_rockylinux8_cabal-build-all.sh.yml/dispatches --data '{"ref": "alpha"}'
```

File: .github/workflows/premerge_ubuntu20_prereqs.sh.yml
Comment (TrevorBenson) [inline .github/workflows/premerge_ubuntu20_prereqs.sh.yml:46]:
```suggestion
        curl -XPOST -u "${{ secrets.REPO_SCOPED_USER}}:${{secrets.REPO_SCOPED_TOKEN }}" -H "Accept: application/vnd.github.everest-preview+json" -H "Content-Type: application/json" https://api.github.com/repos/cardano-community/guild-operators/actions/workflows/premerge_ubuntu20_cabal-build-all.sh.yml/dispatches --data '{"ref": "alpha"}'
```

File: files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/amazonlinux2-cabal-_l.containerfile:15]:
```suggestion
RUN curl -o cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

File: files/tests/pre-merge/amazonlinux2-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/amazonlinux2-cabal.containerfile:15]:
```suggestion
RUN curl -o cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

File: files/tests/pre-merge/debian-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/debian-cabal.containerfile:16]:
```suggestion
RUN curl --silent --show-error --insecure --output cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

File: files/tests/pre-merge/debian-cabal_-l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/debian-cabal_-l.containerfile:16]:
```suggestion
RUN curl --silent --show-error --insecure --output cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

File: files/tests/pre-merge/rockylinux8-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/rockylinux8-cabal.containerfile:15]:
```suggestion
RUN curl -o cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

File: files/tests/pre-merge/rockylinux8-cabal_-l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/rockylinux8-cabal_-l.containerfile:15]:
```suggestion
RUN curl -o cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

File: files/tests/pre-merge/ubuntu20-cabal.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/ubuntu20-cabal.containerfile:16]:
```suggestion
RUN curl --silent --show-error --insecure --output cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

File: files/tests/pre-merge/ubuntu20-cabal_-l.containerfile
Comment (TrevorBenson) [inline files/tests/pre-merge/ubuntu20-cabal_-l.containerfile:16]:
```suggestion
RUN curl --silent --show-error --insecure --output cardano-node-latest.txt https://raw.githubusercontent.com/cardano-community/guild-operators/alpha/files/docker/node/release-versions/cardano-node-latest.txt &&\
```

**Comment by rdlrt** [comment]
> Thinking this through and seeing default on alpha, and that PR's branch from alpha, I'm thinking the three bullets above should point to alpha instead of master. Except maybe `cabal.project.local` which could point to GITHUB_REF_NAME for the PR branch, in case testing new cabal param values within the submitted PR is the intended goal.

But such a change would then only test the PR after it's merged, not before right?

**Comment by TrevorBenson** [comment]
> > Thinking this through and seeing default on alpha, and that PR's branch from alpha, I'm thinking the three bullets above should point to alpha instead of master. Except maybe `cabal.project.local` which could point to GITHUB_REF_NAME for the PR branch, in case testing new cabal param values within the submitted PR is the intended goal.
> 
> But such a change would then only test the PR after it's merged, not before right?

The way I was thinking about was if we use `env.GITHUB_REF_NAME` (fixing this now as I left off the `env.` part of the tag originally) from the PR then we test the cabal.project.local from the PR branch itself, not from alpha. If the author did not modify it since the branching from alpha, we test the same file. If the author modified it however we test their changes during these pre-merge testing and it should isolate any build failures that would normally have only been found once the PR was merged into alpha.

I believe the logic is sound, but please feel free to correct me if you think I overlooked something.

**Comment by TrevorBenson** [comment]
> I believe the logic is sound, but please feel free to correct me if you think I overlooked something.

That said I'm still investigating proper usage of GITHUB_REF_NAME because it seems to be undefined when I test a manual workflow run. So once I get it working I will verify it works as I was intending.

**Comment by rdlrt** [comment]
> from the PR then we test the `cabal.project.local` from the PR branch itself, not from `alpha`. If the author did not modify it since the branching from alpha, we test the same file.

That would be the case if you use env variable, not sure if you intended end (current) state to have references to alpha branch in that case , but I assume you've already thought about it - will wait for your changes to finish 👍🏻

**Comment by TrevorBenson** [comment]
@rdlrt I think it's ready for another look.

**Review by rdlrt — APPROVED** [review]
Thanks, LGTM!

**Comment by TrevorBenson** [comment]
I thought of an image name collision due to the use of ghcr.io. I'll post an additional change to ensure multiple PR's submitted simultaneously won't test from the wrong prerequisites image. Will re-request review.

**Comment by TrevorBenson** [comment]
@rdlrt 
OK Commit 11b0443 should limit collisions. The image names have been reformatted and now include the commit hash. This should speed and simplify investigation of build issues as we can not pull a prerequisite image of a specific commit and reproduce the build failure locally.

The only collision that remains is within a single PR branch. If a PR branch has two commits within less than 10 minutes of each other which complies with the trigger. When the first commits prerequisites completes and the build job kicks off it will checkout the second commit and find its commit hash instead. This will cause a failure in the first pre-merge tests build job, as the image won't yet exist from the second commit. 

I don't consider this a big issue. Especially because if someone commits multiple times quickly to the single PR branch it would be courteous as well as "green" to simply stop the earlier unneeded running jobs.

**Review by rdlrt — APPROVED** [review]

**Pull request merged**
